### PR TITLE
update(node): AsyncLocalStorage in v12.17

### LINKED ISSUE #45103: Update for Node v12.17
If you know how to fix the issue, make a pull request instead.

- [X] I tried using the `@types/xxxx` package and had problems.
- [X] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [X] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [X] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @SimonSchick @jkomyno @a-tarasyuk @alvis @r3nya @btoueg @BrunoScheufler @smac89 @tellnes @Touffy @DeividasBakanas @eyqs @Flarna @Hannes-Magnusson-CK @KSXGitHub @hoo29 @kjin @ajafff @islishude @mwiktorczyk @mohsen1 @n-e @octo-sniffle @galkin @parambirs @eps1lon @SimonSchick @ThomasdenH @WilcoBakker @wwwy3y3 @samuela @kuehlein @j-oliveras @bhongy @chyzwar @trivikr @nguymin4 @yoursunny @qwelias @ExE-Boss @Ryan-Willpower @peterblazejewicz (copied the list from https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44097)

Node v12.17 backported `AsyncLocalStorage` in `async_hooks` from v14. Please update @types/node to enable using it in Typescript.

Thanks.

### PR DESCRIPTION
**Pull request by peterblazejewicz**
As per #45103 request, this ports AsyncLocalStorage to the v12

- definition file added to v12
- tests added

https://github.com/nodejs/node/releases/tag/v12.17.0

/cc @orgads

Thanks!

Closes #45103

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes

### CODE REVIEW

File: types/node/v12/async_hooks.d.ts [hunk 1]
```diff
@@ -7,6 +7,20 @@ declare module "async_hooks" {
...
+     *
+     * Resource objects returned by `executionAsyncResource()` are most often internal
+     * Node.js handle objects with undocumented APIs. Using any functions or properties
+     * on the object is likely to crash your application and should be avoided.
+     *
+     * Using `executionAsyncResource()` in the top-level execution context will
+     * return an empty object as there is no handle or request object to use,
+     * but having an object representing the top-level can be helpful.
+     */
+    function executionAsyncResource(): object;
```
Comment (peterblazejewicz) [inline types/node/v12/async_hooks.d.ts:22]: https://nodejs.org/dist/latest-v12.x/docs/api/async_hooks.html#async_hooks_async_hooks_executionasyncresource

File: types/node/v12/async_hooks.d.ts [hunk 2]
```diff
@@ -20,7 +34,7 @@ declare module "async_hooks" {
          * @param triggerAsyncId the unique ID of the async resource in whose execution context this async resource was created
          * @param resource reference to the resource representing the async operation, needs to be released during destroy
          */
-        init?(asyncId: number, type: string, triggerAsyncId: number, resource: Object): void;
+        init?(asyncId: number, type: string, triggerAsyncId: number, resource: object): void;
```
Comment (peterblazejewicz) [inline types/node/v12/async_hooks.d.ts:37]: changed to match:
https://github.com/peterblazejewicz/DefinitelyTyped/blob/feat/45103/types/node/async_hooks.d.ts#L37

**Comment by peterblazejewicz** [comment]
I'm seeing the same tests now failing locally when based on up-to-date master. Will ask on Gitter about the gibberish tests.

**Comment by orgads** [comment]
ping?

**Comment by orgads** [comment]
Anyone up for approval?

**Review by Flarna — APPROVED** [review]
Looks fine except the nit to remove duplicates in tests.

File: types/node/v12/test/async_hooks.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
+import { AsyncResource, createHook, triggerAsyncId, executionAsyncId, executionAsyncResource, HookCallbacks, AsyncLocalStorage } from 'async_hooks';
```
[the remark this answers is not in the document]
Comment (peterblazejewicz) [reply types/node/v12/test/async_hooks.ts:1]: @Flarna amended with 14e1c611d8
Thanks for the review!

**Review by Flarna — APPROVED** [review]

**Review by galkin — APPROVED** [review]

**Review by a-tarasyuk — APPROVED** [review]

**Comment by orta** [comment]
Alright, this looks like enough approvals to me, thank you folk

**Pull request merged**